### PR TITLE
[WIP] Add a Unit Test to test if the vega-lite-schema.json is up to date #1773

### PR DESCRIPTION
**Pull request by nung25649**
I just started working on this pull request and I am stuck on how to proceed and what is the right way to fix the issue. So far I have added the `npm run schema`, but I'm not sure how I should make the test fail if the schema file is different than the committed one. Do I add another test case inside `schema.test.ts` or write a shell script to execute when the `npm run test` is called to check git status?

### CODE REVIEW

**Comment by domoritz** [comment]
I think we need a shell script or simply another script in `package.json` that checks for git changes and returns an error status when the schema has been changed. I believe we don't want to call this test every single time we run `npm run test`, though because this would be annoying when we develop locally.

**Comment by kanitw** [comment]
@nung25649  is this finished or just a WIP? If WIP, add [WIP] to the title, so I'm not confused.  Thanks!

**Comment by nung25649** [comment]
Yes it is! Sorry I forgot that. 

One more question: How do we make the travis test throw an error if our bash file says that the committed schema doesn't match our current version of schema? Is it okay to just `echo "schema doesn't match` in our bash file so that the output appears when we run test?

**Comment by domoritz** [comment]
@nung25649 I believe if the script returns a [unix exist status != 0](http://bencane.com/2014/09/02/understanding-exit-codes-and-how-to-use-them-in-bash-scripts/), builds will fail on travis.

**Comment by nung25649** [comment]
I just finished writing a script. Please check if I skipped or missed something as I'm pretty new to bash shell script.

File: .travis.yml [hunk 1]
```diff
@@ -5,6 +5,7 @@ node_js:
   - "7"
 script:
   - npm run build
+  - npm run schema
```
Comment (kanitw) [inline .travis.yml]: Not sure what you need to add run schema -- build already run schema

Comment (kanitw) [inline .travis.yml]: Not sure why you need to add run schema -- build already run schema

**Comment by kanitw** [comment]
You haven't run the script you wrote from travis at all, so it's definitely unfinished.  

What you should try is:

1) Write a proper script that travis will call.  Then push, your first commit should pass the test. 
2) Add another commit to modify an interface of Vega-Lite (e.g. add string "aaa" as a new valid string literal to `AggregateOp`) and test if you push the code (without the updated schema) to this PR and travis should fail in this second commit. 

After you successfully do this, we can merge your code but ignore the commit that you add to demonstrate that test will fail in a desired situation.

File: scripts/check-schema.sh [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+npm run schema
```
Comment (kanitw) [inline scripts/check-schema.sh]: build already run schema, so you definitely don't need to run twice

File: test/schema.test.ts [hunk 3]
```diff
@@ -22,4 +22,6 @@ describe('Schema', function() {
     }
     assert.equal(valid, true);
   });
+
+  // write another test case here to check that the schema file is different than the committed one
```
Comment (kanitw) [inline test/schema.test.ts]: As @domoritz mentioned, you probably don't have to do this -- unless I misunderstand something.

**Comment by nung25649** [comment]
Okay, I misunderstood the issue. I thought that travis test is run when call `npm run test`, but it is actually the test that runs on github right?

If so, then how can I tell travis to run the script? Do I add that in `.travis.yml`?

**Comment by kanitw** [comment]
> If so, then how can I tell travis to run the script? Do I add that in .travis.yml?
Yes.  For more information, look at Travis documentation online.

File: .travis.yml [hunk 4]
```diff
@@ -6,7 +6,7 @@ node_js:
 script:
   - npm run build
   - npm run schema
-  - npm run test:noschema # already run schema in build
+  - npm run test:noschema && npm run checkschema # already run schema in build
```
Comment (domoritz) [inline .travis.yml]: You can have the commands in separate lines.

**Comment by nung25649** [comment]
Hmm, the scripts work on my local when I changed the schema file and run the script. However, when I test the script on travis by modifying one of the source file and then push, it didn't fail the test as I expected. Is there any specific details about travis that I might not know?

**Comment by kanitw** [comment]
@nung25649 – How about you try to push a commit that suppose to fail and open travis's test log.  That might give you a clue what's wrong.

**Comment by kanitw** [comment]
@nung25649 Can you add a commit that suppose to fail to demonstrate that your script will fail in that case? Otherwise I can't know if it's working correctly.  (If it looks correctly, I'll then merge your stuff ignoring the failure test commit.)

**Comment by kanitw** [comment]
@nung25649  Sweet.  The test looks looks correct now. 

1. Looks like this PR accidentally brings back `namespace AggregateOp` which seems unnecessary -- not sure if you have noticed that. (In general, always review all your changes on github as if you are a reviewer first before asking us to review -- this will help us reduce obvious things to fix and improve review turnaround time. 😄  )

I believe that the root cause is from resolving conflicts incorrectly.  


2.  Since this is on your repo (not the main Vega-Lite repo), it's harder for me to do selective merge myself.  Can you do interactive rebase to drop the "test fail case" and "merge branch ..." commits and force push?  (If you haven't, this is a good time to learn interactive rebase.) 

To summarize, you should do interactive rebase to
1) drop the "test fail case" and "merge branch ..." commits


2. since we are using rebase workflow, in general you should not merge master into your branch.

**Comment by kanitw** [comment]
(A few corrections -- I accidentally hit comment before finish drafting)

@nung25649  Sweet.  The test looks looks correct now.   However:

1. Looks like this PR accidentally brings back `namespace AggregateOp` which seems unnecessary -- not sure if you have noticed that. (In general, always review all your changes on github as if you are a reviewer first before asking us to review -- this will help us reduce obvious things to fix and improve review turnaround time. 😄  )

I believe that the root cause is from forking from older master (older than we merge your previous PR that remove AggregateOp).

2.  Since this is on your repo (not the main Vega-Lite repo), it's harder for me to do selective merge myself.  Can you do interactive rebase to drop the "test fail case" and "merge branch ..." commits and force push?  
(Since we are using rebase workflow, in general you should not merge master into your branch.)

To summarize, you should do interactive rebase to
1) rebase to the latest `origin/master`  (`git fetch` then `git rebase -i origin master`)
2) resolve conflicts correctly (don't bring `namespace AggregateOp` back).
2) drop the "test fail case" and "merge branch ..." commits 

 (If you haven't, this is a good time to learn interactive rebase.)

**Comment by nung25649** [comment]
@kanitw Thanks! I also squashed every commit into one. Please let me know if there's still any changes!

**Comment by kanitw** [comment]
@nung25649  Awesome.  This looks pretty neat now!
